backend/src/composite/mod.rs

```diff
@@ -231,6 +244,120 @@ fn set_size<F: Clone>(pil: Arc<Analyzed<F>>, degree: DegreeType) -> Arc<Analyzed
     }
 }
 
+mod sub_prover {
+    use std::{
+        collections::BTreeMap,
+        sync::{
+            mpsc::{Receiver, SyncSender},
+            Arc, Mutex,
+        },
+        thread::{Scope, ScopedJoinHandle},
+    };
+
+    use crate::{Backend, Error};
+    use powdr_executor::witgen::WitgenCallback;
+    use powdr_number::FieldElement;
+
+    /// Runs a prover until it either completes or challenges the caller, in which case the
+    pub fn run<'s, 'env, 'outer, F: FieldElement>(
+        scope: &'s Scope<'s, 'env>,
+        prover: &'env Mutex<Box<dyn Backend<'outer, F> + 'outer>>,
+        witness: Vec<(String, Vec<F>)>,
+    ) -> RunStatus<'s, F>
+    where
+        'outer: 'env,
+    {
+        let (challenge_sender, challenge_receiver) = std::sync::mpsc::sync_channel(0);
+        let (response_sender, response_receiver) = std::sync::mpsc::sync_channel(0);
+
+        let thread = scope.spawn(move || {
+            // We must ensure the backend can call the callback from
+            // multiple threads (needed by Plonky3, as it requires
+            // WitgenCallback to be Sync). So we wrap its Receiver in Mutex.
+            let response_receiver = Mutex::new(response_receiver);
+
+            let callback = WitgenCallback::<F>::new(Arc::new(move |_, challenge, _| {
+                // Lock first, to tie one challenge to one response.
+                let receiver_guard = response_receiver.lock().unwrap();
```

Comment:
This is to please the compiler, right? We wouldn't expect this to be called from multiple threads? I think it would break the algorithm, because we'd be sending challenges multiple times in the same phase from the same machine?

If that's true, Is there a way we can ensure this doesn't happen?

Reply:
This is related a few lines above, where I place the `response_receiver` inside a mutex. I do it because the `WitgenCallback` object requires it to be `Sync`. And the `WitgenCallback` has to be `Sync` otherwise compilation fails inside plonky3. So I guess plonky3 passes references to this object around threads in its internal implementation. But does it actually calls it from multiple threads? I don't know.

Anyway, the line you tagged is precisely so that the algorithm will still work if callback is called multiple times from the same prover in different threads. This line will lock the mutex for the rest of the function, so whatever thread called first will be seen as a first stage, and any other thread that calls it meanwhile will be blocked until this stage is finished.

Maybe I should `assert!()` that the `stage` argument is being provided sequentially, in the expected order?

Reply:
Or maybe plonky3 is wrong and `WitgenCallback` should not need to be part of a struct that is `Sync`.